### ISSUE TITLE
ttnn.sort: CrossCoreDataExchange ROW_MAJOR inputs with >32 rows return wrong values for every tile-row block after the first

### ISSUE THREAD

**Issue by iwroszTT** [tenstorrent/tt-metal#54309]
Repro (BH p100a, also expected on any grid routing W=4096 to CrossCore): sort a (64, 4096) bf16 ROW_MAJOR tensor, dim=-1, either direction, either stability — rows 32–63 of the returned values do not match torch.sort (per-row max ATOL ~2.5 on a 6-level tie-heavy input; row 32 alone fails for (33, 4096)). Indices remain a valid per-row permutation and TILE layout is unaffected, so the defect is in the RM value transport (reader/writer paging of tile-row blocks ≥ 2), not the sort network. Pre-existing on main: RM test coverage stops at single-core widths and the multicore RM cells are uint16 (which never routes to CrossCore). A non-strict xfail repro ships in #53557 as test_sort_unstable_row_major_multi_tile_row_values.

```
"""Repro: ttnn.sort on ROW_MAJOR inputs with more than 32 rows returns wrong VALUES
for every tile-row block after the first, on the CrossCoreDataExchange factory.

- W=4096 routes to CrossCore on any grid with >= 64 compute cores.
- TILE layout passes; ROW_MAJOR fails from row 32 on ((33, 4096) fails in row 32
  alone; (64, 4096) fails in all of rows 32..63).
- Indices remain a valid per-row permutation in both layouts — only values go wrong,
  so gather-parity checks do not catch it.
- Reproduces identically with stable=True where available; the sort network is not
  involved. Suspect: RM value transport (reader/writer paging of tile-row blocks >= 2).

Self-contained: run with  pytest test_sort_rm_multi_tile_row_repro.py
"""
import pytest
import torch
import ttnn


@pytest.fixture
def device():
    dev = ttnn.open_device(device_id=0)
    yield dev
    ttnn.close_device(dev)


@pytest.mark.parametrize("layout", [ttnn.Layout.ROW_MAJOR, ttnn.Layout.TILE], ids=["row_major", "tile"])
def test_sort_multi_tile_row_values(layout, device):
    torch.manual_seed(0)
    levels = torch.tensor([-1.5, -0.5, 0.0, 0.5, 1.5], dtype=torch.bfloat16)
    input_tensor = levels[torch.randint(0, len(levels), (64, 4096))]

    torch_values, _ = torch.sort(input_tensor, dim=-1)

    ttnn_input = ttnn.from_torch(input_tensor, ttnn.bfloat16, layout=layout, device=device)
    ttnn_values, ttnn_indices = ttnn.sort(ttnn_input, dim=-1)

    dev_values = ttnn.to_torch(ttnn_values)
    dev_indices = ttnn.to_torch(ttnn_indices).to(torch.int64)

    # Indices are a valid permutation in both layouts — only the values go wrong.
    assert torch.equal(
        torch.sort(dev_indices, dim=-1).values, torch.arange(4096).expand(64, -1)
    ), "indices are not a permutation per row"

    bad_rows = (dev_values != torch_values).any(dim=-1).nonzero().flatten().tolist()
    assert not bad_rows, (
        f"sorted values wrong in {len(bad_rows)} rows: {bad_rows[:8]}{'...' if len(bad_rows) > 8 else ''}"
    )

```
